### ISSUE TITLE
dup/with cannot be used to modify fields?

### ISSUE THREAD

**Issue by bluenote10** [nim-lang/Nim#13967]
### Summary

I was trying to upgrade a DSL of mine to modern Nim using `dup`/`with` (relying on a builder pattern so far), but I cannot figure out how to write one-line-expressions that modify fields.

I thought I can use them to turn an immutable instance of an object into a new immutable instance with some fields modified (the equivalent of case class copies like `original.copy(field="modified")` in Scala).

### Description

```nim
import sugar, with

type
  User = object
    id: int
    name: string
    age: int
  
let user = User(id: 42, name: "original", age: 42)
echo user
echo user.dup(name="modified")
echo user.with(name="modified")
echo user.dup.with(name="modified")
echo user.shallowCopy().with(name="modified")
```

First line errors with:
```
Error: invalid expression: name = "modified"
```

Second / third lines have similar errors:
```
Error: type mismatch: got <User, name: string>
but expected one of: 
macro with(x: typed; body: untyped): untyped
  first type mismatch at position: 2
  unknown named parameter: name

expression: with(
  var dupResult_3786220 = user
  dupResult_3786220, name = "modified")
```

Forth line errors with:
```
Error: type mismatch: got <User>
but expected one of: 
proc shallowCopy[T](x: var T; y: T)
  first type mismatch at position: 1
  required type for x: var T
  but expression 'user' is immutable, not 'var'

expression: shallowCopy(user)
```

The only expression that I got to compile so far was
```nim
echo (block:
  var tmpUser = user
  with tmpUser:
    name = "modified"
  tmpUser)
```
which isn't a real help to replace a builder pattern.

Am I missing something? If not, is field modification a feature that can be added to `with`/`dup`?

**Comment by Araq** [comment]
Field modifications are not part of the design. Can they be added? I don't know, feel free to tinker with the implementation, it's rather simple IMO.

**Comment by Clyybber** [comment]
Fixed by #14484
